include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,454 @@
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_get_supported_fields(dev, mask);
+}
+
+/**
+ * @brief API for setting RTC alarm time
+ *
+ * @details To enable an RTC alarm, one or more fields of the RTC alarm time
+ * must be enabled. The mask designates which fields of the RTC alarm time to
+ * enable. If the mask parameter is 0, the alarm will be disabled. The RTC
+ * alarm will trigger when all enabled fields of the alarm time match the RTC
+ * time.
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm
+ * @param mask Mask of fields in the alarm time to enable
+ * @param timeptr The alarm time to set
+ *
+ * @note The timeptr param may be NULL if the mask param is 0
+ *
+ * @note Only the enabled fields in the timeptr param need to be set
+ *
+ * @note Bits in the mask param are defined in the RTC_ALARM_TIME_MASK defines
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range or time is invalid
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+				 const struct tm *timeptr);
+
+static inline int z_impl_rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+					    const struct tm *timeptr)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_set_time(dev, id, mask, timeptr);
+}
+
+/**
+ * @brief API for getting RTC alarm time
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm
+ * @param mask Destination for mask of fields which are enabled in the alarm time
+ * @param timeptr Destination for the alarm time
+ *
+ * @note bits in mask are stored in RTC_ALARM_TIME_MASK defines
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				 struct tm *timeptr);
+
+static inline int z_impl_rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+					    struct tm *timeptr)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_get_time(dev, id, mask, timeptr);
+}
+
+/**
+ * @brief API for testing if RTC alarm is pending
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm to test
+ *
+ * @return 1 if alarm is pending
+ * @return 0 if alarm is not pending
+ * @return -EINVAL if id is out of range
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_is_pending(const struct device *dev, uint16_t id);
+
+static inline int z_impl_rtc_alarm_is_pending(const struct device *dev, uint16_t id)
```

Comment:
Every RTC I've seen has multiple flags in a single register.  Can we use a bitmask of alarms, so we don't have to re-read the status register for each alarm?  This is how Zephyr counter drivers work for multiple alarms.

Reply:
I prefer keeping the interactions with alarms uniform throughout the API. Since each alarm is an individual object, the API functions mimic this. The pending status is directly tied to each individual alarm.

The RTCs I have worked with have all their interrupt sources in a single register, not just the alarms, "picking out" the specific flag we are interrested in using the id therefore seems like the most user friendly thing to do, at the cost of having to perform to API calls to check two alarms.

Reply:
It seems like you're only considering RTCs that are built into the MCU, like the Atmel SAM in this PR.  For external RTC chips, the many I2C transactions to query each flag one at a time is quite a bit more costly.

While I agree that considering this API alone, using alarm IDs uniformly is conceptually cleanly, one has to also consider Zephyr as a whole.  Zephyr's APIs for counter alarms and GPIO interrupts both use integer identifiers or bitmasks depending on the context: if the method naturally supports a single item or a set of items.  IMHO, it's not cleaner that GPIOs can be queried as a group, and counter alarms can be queried as a group, but RTC alarms can not be.

Reply:
I will update the API to return a bitmask, like the counter and gpio APIs. Since there is precedence for the bitmask / id combo, it is a fine way to make the API more efficient.

I didn't expect applications to poll these bits often, which was my foundation for expecting the transaction cost to be acceptable.

Reply:
Sometimes the RTC IRQ line is not connected.  It takes a pin and pins can be scarce.  The way to test for an RTC alarm would be to poll it in that case, which might be more often.